File: lab2/gr8sub1/Лобанок_Иван/src/fork_example.c

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>   // fork(), getpid(), getppid()
#include <sys/wait.h> // wait()

int main() {
    printf("parent: PID=%d start\n", getpid());
    fflush(stdout);

    for (int i = 0; i < 2; i++) {
        pid_t pid = fork();

        if (pid < 0) {
            perror("fork");
            exit(1);
        }
        else if (pid == 0) {
            printf("child[%d]: PID=%d, PPID=%d\n", i, getpid(), getppid());
            fflush(stdout);
            _exit(0);
        }
   
    }

    int status;
    pid_t wpid;
    while ((wpid = wait(&status)) > 0) {
        if (WIFEXITED(status)) {
            printf("parent: child PID=%d exited, code=%d\n",
                   wpid, WEXITSTATUS(status));
        }
        fflush(stdout);
    }

    printf("parent: PID=%d finished\n", getpid());
    fflush(stdout);

    return 0;
}

File: lab2/gr8sub2/КАЗАК_АРТУР/src/fork_example.py
#!/usr/bin/env python3
import os
import sys

def main() -> int:
    print(f"parent: start PID={os.getpid()}", flush=True)
    children = []

    for i in range(2):
        pid = os.fork()
        if pid == 0:
            # child
            print(f"child[{i}]: PID={os.getpid()}, PPID={os.getppid()}", flush=True)
            os._exit(0)
        else:
            # parent
            children.append(pid)

    # wait for both children, report exit status
    for pid in children:
        wpid, status = os.waitpid(pid, 0)
        if os.WIFEXITED(status):
            code = os.WEXITSTATUS(status)
            print(f"parent: child PID={wpid} exited with code {code}", flush=True)
        elif os.WIFSIGNALED(status):
            print(f"parent: child PID={wpid} killed by signal {os.WTERMSIG(status)}", flush=True)
        else:
            print(f"parent: child PID={wpid} status={status}", flush=True)

    print(f"parent: done PID={os.getpid()}", flush=True)
    return 0

if __name__ == "__main__":
    sys.exit(main())
